compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2552,6 +2554,50 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             _ => {}
         }
     }
+
+    fn suggest_derive(&self, err: &mut Diagnostic, trait_pred: ty::PolyTraitPredicate<'tcx>) {
+        let Some(diagnostic_name) = self.tcx.get_diagnostic_name(trait_pred.def_id()) else {
+            return;
+        };
+        let (adt, substs) = match trait_pred.skip_binder().self_ty().kind() {
+            ty::Adt(adt, substs) if adt.did().is_local() => (adt, substs),
+            _ => return,
+        };
+        // Ensure all fields impl the trait.
+        for variant in adt.variants() {
+            for field in &variant.fields {
+                let obligation = self.mk_trait_obligation_with_new_self_ty(
```

Comment:
So I was wrong here -- you need to fix two things:
1. Actually don't use `tcx.param_env`... you need to get the param-env from the obligation (from the function that calls `suggest_derive`).

2. The substs are not exactly correct for `PartialEq` and `PartialOrd`. Imagine this struct:
```rust
struct Foo {
  bar: String,
}
```
If you have `Foo: PartialEq<Foo>`, then you will currently create a child obligation for the `foo` field that looks like `String: PartialEq<Foo>`, instead of `String: PartialEq<String>`. Instead of using `mk_trait_obligation_with_new_self_ty`, you need to create a new trait ref with a new self type and possibly extra subst (for PartialEq and PartialOrd).

Reply:
Thank you for the details! I understand.
I'll modify it to create a new trait ref like `String: PartialEq<String>`